ante/fee_test.go

```diff
@@ -0,0 +1,579 @@
+	cheqPrice := sdk.NewDecCoinFromDec("cheq", sdk.NewDec(20))
+	highGasPrice := []sdk.DecCoin{cheqPrice}
+	s.ctx = s.ctx.WithMinGasPrices(highGasPrice)
+
+	// Set IsCheckTx to true
+	s.ctx = s.ctx.WithIsCheckTx(true)
+
+	// antehandler errors with insufficient fees
+	_, err = antehandler(s.ctx, tx, false)
+	s.Require().NotNil(err, "Decorator should have errored on too low fee for local gasPrice")
+
+	// antehandler should not error since we do not check minGasPrice in simulation mode
+	cacheCtx, _ := s.ctx.CacheContext()
+	_, err = antehandler(cacheCtx, tx, true)
+	s.Require().Nil(err, "Decorator should not have errored in simulation mode")
+
+	// Set IsCheckTx to false
+	s.ctx = s.ctx.WithIsCheckTx(false)
+
+	// antehandler should not error since we do not check minGasPrice in DeliverTx
+	_, err = antehandler(s.ctx, tx, false)
+	s.Require().Nil(err, "MempoolFeeDecorator returned error in DeliverTx")
+
+	// Set IsCheckTx back to true for testing sufficient mempool fee
+	s.ctx = s.ctx.WithIsCheckTx(true)
+
+	cheqPrice = sdk.NewDecCoinFromDec("cheq", sdk.NewDec(0).Quo(sdk.NewDec(100000)))
+	lowGasPrice := []sdk.DecCoin{cheqPrice}
+	s.ctx = s.ctx.WithMinGasPrices(lowGasPrice)
+
+	newCtx, err := antehandler(s.ctx, tx, false)
+	s.Require().Nil(err, "Decorator should not have errored on fee higher than local gasPrice")
+	// Priority is the smallest gas price amount in any denom. Since we have only 1 gas price
+	// of 10cheq, the priority here is 10.
+	s.Require().Equal(int64(10), newCtx.Priority())
+}
+
+func (s *AnteTestSuite) TestDeductFees() {
+	s.SetupTest(false) // setup
+	s.txBuilder = s.clientCtx.TxConfig.NewTxBuilder()
+
+	// keys and addresses
+	priv1, _, addr1 := testdata.KeyTestPubAddr()
+
+	// msg and signatures
+	msg := testdata.NewTestMsg(addr1)
+	feeAmount := NewTestFeeAmount()
+	gasLimit := testdata.NewTestGasLimit()
+	s.Require().NoError(s.txBuilder.SetMsgs(msg))
+	s.txBuilder.SetFeeAmount(feeAmount)
+	s.txBuilder.SetGasLimit(gasLimit)
+
+	privs, accNums, accSeqs := []cryptotypes.PrivKey{priv1}, []uint64{0}, []uint64{0}
+	tx, err := s.CreateTestTx(privs, accNums, accSeqs, s.ctx.ChainID())
+	s.Require().NoError(err)
+
+	// Set account with insufficient funds
+	acc := s.app.AccountKeeper.NewAccountWithAddress(s.ctx, addr1)
+	s.app.AccountKeeper.SetAccount(s.ctx, acc)
+	coins := sdk.NewCoins(sdk.NewCoin("cheq", sdk.NewInt(10)))
+	//nocheck:errcheck
+	err = testutil.FundAccount(s.app.BankKeeper, s.ctx, addr1, coins)
+	s.Require().NoError(err)
+
+	dfd := cheqdante.NewDeductFeeDecorator(s.app.AccountKeeper, s.app.BankKeeper, nil, nil)
+	antehandler := sdk.ChainAnteDecorators(dfd)
+
+	_, err = antehandler(s.ctx, tx, false)
+
+	s.Require().NotNil(err, "Tx did not error when fee payer had insufficient funds")
+
+	// Set account with sufficient funds
+	s.app.AccountKeeper.SetAccount(s.ctx, acc)
+	err = testutil.FundAccount(s.app.BankKeeper, s.ctx, addr1, sdk.NewCoins(sdk.NewCoin("cheq", sdk.NewInt(200))))
+	s.Require().NoError(err)
```

Comment:
People always provide their fees in base denom in ncheq, so isn't this supposed to be in ncheq>

Reply:
These are native tests in `u` units, so the denom is not relevant as long as the comparison is under the same metric, but for sure it makes sense to keep in the correct line.

The following three are ensuring we're following default fee deduction logic for native SDK tx types:

- `TestDeductFeeDecorator_ZeroGas`
- `TestEnsureMempoolFees`
- `TestDeductFees`

The updated version accounts for the denominations across the whole logic.

I've pushed a